mrmustard/utils/training.py

```diff
@@ -85,10 +90,20 @@ def should_stop(self, max_steps: int) -> bool:
                 sum(abs(self.opt_history[-i - 1] - self.opt_history[-i]) for i in range(1, 20))
                 < 1e-6
             ):
-                print("Loss looks stable, stopping here.")
+                self.log.info("Loss looks stable, stopping here.")
                 return True
         return False
 
+    class OptimizerInterruptedError(Exception):
+        """A helper class to quietly stop execution without printing a traceback."""
+
+        def __init__(self, message: str, logger: Logger) -> None:
+            super().__init__()
+            logger.info(message)
```

Comment:
What's the reason to put the logging inside of the exception initialization? Would it perhaps make more sense to have the logging where the exception is raised instead?

Reply:
No good reason indeed. It does make more sense to have the logging where you suggest 🤓